aten/src/THC/THCTensorMath.cuh

```diff
@@ -95,35 +77,44 @@ struct OutputTensorSizeStride {
   *
   * The most important assumption made is that the input tensors are contiguous.
   */
-
-
-
 template <typename T, typename IndexType, int Dims>
 __global__ void CatArrayBatchedCopy(
     T* output,
-    CatArrInputTensor<T, IndexType>* inputs,
+    CatArrInputTensor<T, IndexType, CAT_ARRAY_MAX_INPUT_DIMS>* inputs,
     OutputTensorSizeStride<IndexType, CAT_ARRAY_MAX_INPUT_DIMS> os,
     const int concatDim,
-    IndexType dimStride) {
-
-    IndexType tid = blockIdx.x * blockDim.x + threadIdx.x;
-    IndexType nElements = inputs[blockIdx.y].nElements;
+    T pad_value) {
 
-    if(tid >= nElements) return;
-    
+    IndexType nElementsOutput = inputs[blockIdx.y].nElementsOutput;
     T* data = inputs[blockIdx.y].input;
-    IndexType offset = inputs[blockIdx.y].offset;
-    IndexType dimSize = inputs[blockIdx.y].dimSize;
-    IndexType dataOffset = offset * dimStride;
-
-    IndexType stride = gridDim.x * blockDim.x;
-
-    while( tid < nElements){
-    IndexType elementOffset = CatArrIndexToOffset<IndexType, Dims>::compute(
-    	      os.outputSize, os.outputStride, dimSize, concatDim, tid);
-    output[dataOffset + elementOffset] = data[tid];
-
-    tid += stride;
+    IndexType dimOffset = inputs[blockIdx.y].offset;
+    IndexType dataOffset = dimOffset * os.outputStride[concatDim];
+
+    for (IndexType linearIndex = (IndexType) blockIdx.x * blockDim.x + threadIdx.x;
+      linearIndex < nElementsOutput;
+      linearIndex += (IndexType) gridDim.x * blockDim.x) {
+      IndexType inputOffset = 0;
+      IndexType outputOffset = 0;
+      bool inbound = true;
+      for (int i = Dims - 1; i >= 1; --i) {
+        IndexType inputDimSize = inputs[blockIdx.y].inputParam.inputSize[i];
+        IndexType curDimSize = i == concatDim ? inputDimSize : os.outputSize[i];
+        IndexType nextDimIndex = linearIndex / curDimSize;
+        IndexType curDimIndex = linearIndex - curDimSize * nextDimIndex;
+        if (curDimIndex >= inputDimSize && inbound) {
+          inbound = false;
+        }
+        inputOffset += curDimIndex * inputs[blockIdx.y].inputParam.inputStride[i];
+        outputOffset += curDimIndex * os.outputStride[i];
+        linearIndex = nextDimIndex;
+      }
+      inputOffset += linearIndex * inputs[blockIdx.y].inputParam.inputStride[0];
```

Comment:
Should you be checking that linearIndex is less than inputSize[0], because you might be padding 0-th dimension?